linera-execution/src/execution.rs

```diff
@@ -189,30 +195,30 @@ where
         application_id: UserApplicationId,
         chain_id: ChainId,
         action: UserAction,
-        policy: &ResourceControlPolicy,
-        tracker: &mut ResourceTracker,
+        resource_controller: &mut ResourceController<Option<Owner>>,
     ) -> Result<Vec<ExecutionOutcome>, ExecutionError> {
-        let balance = self.system.balance.get();
-        let runtime_limits = tracker.limits(policy, balance);
-        let initial_remaining_fuel = policy.remaining_fuel(*balance);
+        let controller = ResourceController {
+            policy: resource_controller.policy.clone(),
+            tracker: resource_controller.tracker,
+            account: resource_controller.with(self).await?.balance(),
+        };
         let (execution_state_sender, mut execution_state_receiver) =
             futures::channel::mpsc::unbounded();
         let execution_outcomes_future = tokio::task::spawn_blocking(move || {
             ContractSyncRuntime::run_action(
                 execution_state_sender,
                 application_id,
                 chain_id,
-                runtime_limits,
-                initial_remaining_fuel,
+                controller,
                 action,
             )
         });
         while let Some(request) = execution_state_receiver.next().await {
             self.handle_request(request).await?;
```

Comment:
None of these can change the owner's balance, can they?

Reply:
Not yet

Reply:
If they can, won't that be a problem, since we just reset the balance later to the number from the controller?

Reply:
If we're talking about the same thing, I saw that and made a fix is in the next PR: #1548